File: server/score.rb
#!/usr/bin/ruby -w


# Public: Called before a match to calculate the expected score
#
# _p1_cur_score - Current score of player making request
# _p2_cur_score - Current score of opponent
#
# Returns expected score of player
#
def prematchScores(_p1_cur_score, _p2_cur_score)
	# Calculate transformed scores
	tp1 = 10**(_p1_cur_score / 400)
	tp2 = 10**(_p2_cur_score / 400)

	# Return expected scores
	return (tp1 / (tp1 + tp2))
end


# Public: Called after a match to calculate the new score
#
# _old_score - Players current score
# _exp_score - The expected score calculated in prematchScores()
# _games_played - The number of games the player has played
# _result - Result of the game (Win - 1, Tie - 0.5, Loss - 0)
#
# Returns the player's new score
#
def postmatchScores(_old_score, _exp_score, _games_played, _result)
	# Calculate k value
	if _games_played < 30 then
		k = 40
	elsif _old_score < 2400 then
		k = 20
	else
		k = 10
	end
	
	# Return new score
	return (_old_score + k * (_result - _exp_score))
end

File: js/score.js
/* Called before a match begins

	myCurScore - The current score of the player calling the function
	oppCurScore - The current score of the opponent
	
	Returns the "expected score" of the player making the call

	**** BOTH PLAYERS MUST CALL THIS FUNCTION TO GET EACH OF THEIR
	EXPECTED SCORES BEFORE THE MATCH BEGINS ****
*/

function prematchScores(myCurScore, oppCurScore) {
	// Calculate the transformed scores
	var tp1 = Math.pow(10, (myCurScore / 400));
	var tp2 = Math.pow(10, (oppCurScore / 400));

	// Calculate the expected score
	var myExpScore = (tp1 / (tp1 + tp2));

	// Return the expected score
	return myExpScore;
}


/* Called after a match ends

	myCurScore - The current score of the player calling the function
	myExpScore - The expected score of the player calling the function
	myGamesPlayed - The number of games that the player has played
	outcome - Result of the game (Win = 1, Tie = 0.5, Loss = 0)

	Returns the player's new score

	**** BOTH PLAYERS MUST CALL THIS FUNCTION TO GET EACH OF THEIR NEW
	SCORES AFTER THE MATCH HAS ENDED ****
*/

function postmatchScores(myCurScore, myExpScore, myGamesPlayed, outcome) {
	// Calculate the k value
	var k = 10;
	if (myGamesPlayed < 30) {
		k = 40;
	} else if (myCurScore < 2400) {
		k = 20;
	}

	// Calculate the new score
	var newScore = (myCurScore + k * (outcome - myExpScore));

	// Return new score
	return newScore;
}
